### PR TITLE
tsnet: add ListenFunnel

### PR DESCRIPTION
**Pull request by crawshaw**
This lets a tsnet binary share a server out over Tailscale Funnel.

### CODE REVIEW

File: tsnet/example/tsnet-funnel/tsnet-funnel.go [hunk 1]
```diff
@@ -2,130 +2,40 @@
...
-
 func main() {
 	flag.Parse()
-	s := new(tsnet.Server)
-	defer s.Close()
-	ctx := context.Background()
-	if err := enableFunnel(ctx, s); err != nil {
-		log.Fatal(err)
+	s := &tsnet.Server{
+		Dir:      "./funnel-demo-config.state",
```
Comment (bradfitz) [inline tsnet/example/tsnet-funnel/tsnet-funnel.go:25]: That looks like a file, not a dir

**Comment by crawshaw** [comment]
Shayne wants to make ListenFunnel do both funnel and tailnet listening, which this PR doesn't do. Investigating.

File: tsnet/tsnet.go [hunk 2]
```diff
@@ -760,6 +761,74 @@ func (s *Server) APIClient() (*tailscale.Client, error) {
 // Listen announces only on the Tailscale network.
 // It will start the server if it has not been started yet.
 func (s *Server) Listen(network, addr string) (net.Listener, error) {
+	return s.listen(network, addr, false)
+}
+
+// ListenFunnel announces on the public internet using Tailscale Funnel.
+//
+// Currently, Funnel only supports TCP on ports 443, 8443, and 10000.
```
Comment (bradfitz) [inline tsnet/tsnet.go:769]: Temporal references should include a date in parens

File: tsnet/tsnet.go [hunk 3]
```diff
@@ -760,6 +761,74 @@ func (s *Server) APIClient() (*tailscale.Client, error) {
...
+// The support host name is limited to that configured for the tsnet.Server.
```
Comment (bradfitz) [inline tsnet/tsnet.go:770]: Supported

File: tsnet/tsnet.go [hunk 4]
```diff
@@ -760,6 +761,74 @@ func (s *Server) APIClient() (*tailscale.Client, error) {
...
+//
+// and the only other supported addrs currently are ":8443" and ":10000".
+//
+// It will start the server if it has not been started yet.
+func (s *Server) ListenFunnel(network string, addr string) (net.Listener, error) {
+	if network != "tcp" {
+		return nil, fmt.Errorf("ListenFunnel(%q, %q): only tcp is supported", network, addr)
+	}
+	switch addr {
+	case ":443", ":8443", ":10000":
```
Comment (bradfitz) [inline tsnet/tsnet.go:783]: Don't hard code this. Boxes us in later

File: tsnet/tsnet.go [hunk 5]
```diff
@@ -760,6 +761,74 @@ func (s *Server) APIClient() (*tailscale.Client, error) {
...
+		return nil, errors.New("tsnet: you must enable HTTPS in the admin panel to proceed")
+	}
+
+	lc, err := s.LocalClient() // do local client first before listening.
+	if err != nil {
+		return nil, err
+	}
+
+	// May not have funnel enabled. Enable it.
+	srvConfig, err := lc.GetServeConfig(ctx)
```
Comment (maisem) [inline tsnet/tsnet.go:802]: Add a check for the funnel nodeAttr too

**Comment by crawshaw** [comment]
I still have to address all the comments, but I believe ListenFunnel now listens on both by default.

File: tsnet/tsnet.go [hunk 6]
```diff
@@ -760,6 +760,95 @@ func (s *Server) APIClient() (*tailscale.Client, error) {
 // Listen announces only on the Tailscale network.
 // It will start the server if it has not been started yet.
 func (s *Server) Listen(network, addr string) (net.Listener, error) {
+	return s.listen(network, addr, listenOnTailnet)
+}
+
+// FunnelOption is an option passed to ListenFunnel to configure the listener.
+type FunnelOption string
```
Comment (maisem) [inline tsnet/tsnet.go:767]: I'm concerned that `string` is going to corner us in the future.

File: tsnet/tsnet.go [hunk 7]
```diff
@@ -760,6 +761,96 @@ func (s *Server) APIClient() (*tailscale.Client, error) {
...
+	}
+	domain := st.CertDomains[0]
+	hp := ipn.HostPort(domain + addr) // valid only because of the strong restrictions on addr above
+	if !srvConfig.AllowFunnel[hp] {
+		mak.Set(&srvConfig.AllowFunnel, hp, true)
+		srvConfig.AllowFunnel[hp] = true
+		if err := lc.SetServeConfig(ctx, srvConfig); err != nil {
+			return nil, err
+		}
+	}
```
Comment (shayne) [inline tsnet/tsnet.go:827]: We should clear the `serveConfig` when a tsnet app starts. However, we're persisting the `serveConfig`, so if the user turns on Funnel with `ListenFunnel` and then changes it to `Listen`, well, it is still on. There are other oddities too.

I would suggest just clobbering it and not looking up the existing config, but a complex tsnet app could be running multiple listeners.

Comment (crawshaw) [reply tsnet/tsnet.go:857]: OK, we could clobber it in Up.

File: tsnet/tsnet.go (hunk 6 above)
Comment (crawshaw) [reply tsnet/tsnet.go:798]: I don't really see how, but what alternative would you prefer?

Comment (maisem) [reply tsnet/tsnet.go:798]: My concern only concern is that we won't be able to pass parameters.

We can do the usual modifier function option thing.

Like

`type ListenOption func(*listenConfig)`

Comment (crawshaw) [reply tsnet/tsnet.go]: I went with the most generic of the option type variants, figuring it gives us the flexibility to convert later.

Comment (maisem) [reply tsnet/tsnet.go]: Thanks!

**Review by maisem — APPROVED** [review]

File: tsnet/tsnet.go (hunk 4 above)
Comment (maisem) [reply tsnet/tsnet.go:830]: I'm working on a separate change for this.

**Pull request merged**